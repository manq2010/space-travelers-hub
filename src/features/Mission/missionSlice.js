/* eslint-disable no-param-reassign */
// Import createSlice() from Redux toolkit:
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from '../../axios';

// Initial state for Redux store:
const initialState = {
  missions: [],
  isLoading: true,
  isMember: true,
  status: 'idle',
  error: '',
};

export const fetchMissions = createAsyncThunk(
  'missions/fetchMissions',
  async () => {
    const response = await axios.get('/missions');
    const missionsArray = Object.keys(response.data).map((key) => ({
      mission_id: key,
      ...response.data[key],
    }));
    const missions = missionsArray.map((mission) => ({
      reserved: false,
      mission_id: mission.mission_id,
      mission_name: mission.mission_name,
      description: mission.description,
    }));

    return missions.slice().sort((a, b) => a.mission_id - b.mission_id);
  },
);

// Create Redux state slice
const missionSlice = createSlice({
  name: 'missions',
  initialState, // Define initial state
  reducers: {
    // Define reducers
    joinMission: (state, action) => {
      state.missions.map((mission) => (mission.mission_id !== action.payload
        ? mission : { ...mission, reserved: true }));
    },
    leaveMission: (state, action) => {
      state.missions.map((mission) => (mission.mission_id !== action.payload
        ? mission : { ...mission, reserved: false }));
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchMissions.pending, (state) => {
        state.isLoading = true;
        state.status = 'loading';
      })
      .addCase(fetchMissions.fulfilled, (state, action) => {
        state.isLoading = false;
        state.missions = action.payload;
        state.status = 'succeeded';
      })
      .addCase(fetchMissions.rejected, (state, action) => {
        state.isLoading = false;
        state.status = 'failed';
        state.error = action.error.message;
      });
  },
});

// Export actions generated by "createSlice()":
export const { membership, joinMission, leaveMission } = missionSlice.actions;

// Export reducer generated by "createSlice()":
export default missionSlice.reducer;
