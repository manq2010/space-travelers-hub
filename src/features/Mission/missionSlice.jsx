/* eslint-disable no-param-reassign */
// Import createSlice() from Redux toolkit:
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from '../../axios';

// Initial state for Redux store:
const initialState = {
  missions: [],
  isLoading: true,
  isMember: true,
  status: 'idle',
  error: '',
};

export const fetchMissions = createAsyncThunk(
  'missions/fetchMissions',
  async () => {
    const response = await axios.get('/missions');
    const missions = Object.keys(response.data).map((key) => ({
      mission_id: key,
      ...response.data[key],
    }));
    return missions.slice().sort((a, b) => a.mission_id - b.mission_id);
  },
);

// Create Redux state slice
const missionSlice = createSlice({
  name: 'missions',
  initialState, // Define initial state
  reducers: {
    // Define reducers
    // addMission: (state) => {
    //   state.missions = [...state.missions];
    // },
    membership: (state) => {
      state.isMember = !state.isMember;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchMissions.pending, (state) => {
        state.isLoading = true;
        state.status = 'loading';
      })
      .addCase(fetchMissions.fulfilled, (state, action) => {
        state.isLoading = false;
        state.missions = action.payload;
        state.status = 'succeeded';
      })
      .addCase(fetchMissions.rejected, (state, action) => {
        state.isLoading = false;
        state.status = 'failed';
        state.error = action.error.message;
      });
  },
});

// Export actions generated by "createSlice()":
export const { membership } = missionSlice.actions;

// Export reducer generated by "createSlice()":
export default missionSlice.reducer;
