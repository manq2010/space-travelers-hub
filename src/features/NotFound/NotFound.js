import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

const NotFoundContainer = styled.div`
  display: flex;
  margin-top: 10rem;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;
const NotFound = () => (
  <NotFoundContainer>
    <h3>404!</h3>
    <p>
      Oops Page
      {' '}
      <em>Not</em>
      {' '}
      Found!
    </p>
    <Link to="/">Back to home</Link>
  </NotFoundContainer>
);

export default NotFound;
