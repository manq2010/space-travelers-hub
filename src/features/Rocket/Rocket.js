import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRocket } from './RocketSlice';

const Rockets = () => {
  const rockets = useSelector((state) => state.rocketReducer);
  // console.log(rockets);

  const dispatch = useDispatch();

  dispatch(fetchRocket());

  return (
    <div>
      {rockets.map((rocket) => (
        <div key={rocket.id}>
          <img src={rocket.flickrImages} alt="rocket" />
          <div>
            <h1>{rocket.name}</h1>
            <p>{rocket.description}</p>
            <button type="button">Reserve</button>
          </div>
        </div>
      ))}
    </div>
  );
};

export default Rockets;
