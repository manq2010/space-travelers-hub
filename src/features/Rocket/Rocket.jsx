import React from "react";

const Rocket = () => {
    return (
        <div>Rockets</div>
    )
}

export default Rocket;