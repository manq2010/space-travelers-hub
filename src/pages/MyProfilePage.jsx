import React from 'react';
import Profile from '../features/Profile/Profile';

const MyProfilePage = () => (
  <div>
    <Profile />
  </div>
);

export default MyProfilePage;
