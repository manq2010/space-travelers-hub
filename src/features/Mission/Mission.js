import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
// Import membership reducer:
import styled from 'styled-components';
import { joinMission, leaveMission, fetchMissions } from './missionSlice';

const MissionWrapper = styled.div`

padding: 1rem;

`;

const MissionUL = styled.ul`
margin: 0 2rem 0 1rem;

`;

const MissionList = styled.li`
list-style: none;
display: grid;
grid-template-columns: 150px 2fr 150px 100px;
margin: 0;
// border: 1px solid gray;

`;

const Header = styled.h4`
display: flex;
justify-content: space-between;
align-items: center;
`;

const MissionName = styled(Header)`

`;

const HeaderDescription = styled(Header)`
display: flex;
flex-direction: column;
`;

const InnerDescription = styled.span`
display: flex;
flex-direction: column;
`;

const StatusButton = styled.div`
display: flex;
justify-content: center;
align-items: center;
`;

const StatusBtnContainer = styled(StatusButton)`
display: flex;
justify-content: center;
align-items: center;
border-right: 2px solid gray;
`;

const ButtonMember = styled.button`
text-transform: uppercase;
`;

const ButtonJoin = styled.button`

`;

const MissionDescription = styled.p`
border-left: 2px solid gray;
padding: 0 10px 0 10px;
font-size: 12px;
display: flex;
border-right: 2px solid gray;
`;

const Mission = () => {
  const { missions } = useSelector((state) => state.missionReducer);
  const [showMore, SetShowMore] = useState(true);

  // console.log(isMember);
  // console.log(membership);
  // // console.log(filterMembership);
  // console.log(status);

  const dispatch = useDispatch();

  // const dispatch = useDispatch();
  // dispatch.dispatch(filterMembership('9D1B7E0'));

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  // useEffect(() => {
  //   if (status === 'idle') {
  //     dispatch(fetchMissions());
  //   }
  // }, [status, dispatch]);

  // if (missions.discription.length >= 200) {
  //   content = missions.discription.substring(0, 200);
  // } else {
  //   content = missions.discription;
  // }

  return (
    <MissionWrapper>

      <MissionUL>
        <MissionList>
          <MissionName>Mission</MissionName>
          <HeaderDescription>
            <span>
              Description
            </span>
            <span>
              {
              showMore ? (
                <button type="button" onClick={() => SetShowMore(!showMore)}>showMore</button>
              ) : (
                <button type="button" onClick={() => SetShowMore(!showMore)}>showLess</button>
              )
            }
            </span>
          </HeaderDescription>
          <Header>Status</Header>
          <Header />
        </MissionList>
        {
            missions.map((mission) => (
              <MissionList key={mission.mission_id}>
                <MissionName>
                  {mission.mission_name}
                </MissionName>
                <MissionDescription>
                  {showMore ? (
                    <InnerDescription>
                      {mission.description.substring(0, 100)}
                      ...
                    </InnerDescription>
                  ) : (
                    <InnerDescription>
                      {mission.description}
                    </InnerDescription>
                  )}

                </MissionDescription>
                <StatusBtnContainer>
                  {
                    mission.reserved ? (
                      <ButtonMember type="button">Not a Member</ButtonMember>
                    ) : (
                      <ButtonMember
                        type="button"
                        onClick={() => {
                          dispatch(leaveMission(mission.mission_id));
                        }}
                      >
                        Active Member
                      </ButtonMember>
                    )
                  }
                </StatusBtnContainer>
                <StatusBtnContainer>
                  {
                    mission.reserved ? (

                      <ButtonJoin type="button">Leave Mission</ButtonJoin>
                    ) : (
                      <ButtonJoin
                        type="button"
                        onClick={() => {
                          dispatch(joinMission(mission.mission_id));
                        }}
                      >
                        Join Mission
                      </ButtonJoin>
                    )
                  }
                </StatusBtnContainer>
              </MissionList>
            ))
        }
      </MissionUL>

    </MissionWrapper>
  );
};

export default Mission;
