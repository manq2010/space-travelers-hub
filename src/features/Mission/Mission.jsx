import React from 'react';
import { useSelector } from 'react-redux';
// Import membership reducer:
import { membership } from './missionSlice';

const Mission = () => {
  const { isMember, missions, status } = useSelector((state) => state.missionReducer);

  console.log(isMember);
  console.log(membership);
  console.log(status);

  //   const dispatch = useDispatch();

  //   useEffect(() => {
  //     dispatch(fetchMissions());
  //   }, [dispatch]);

  //   useEffect(() => {
  //     if (status === 'idle') {
  //       dispatch(fetchMissions());
  //     }
  //   }, [status, dispatch]);

  return (
    <div>

      <ul>
        <li>
          <h4>Mission</h4>
          <h4>Description</h4>
          <h4>Status</h4>
          <></>
        </li>
        {
            missions.map((mission) => (
              <li key={mission.mission_id}>
                <h4>
                  {mission.mission_name}
                </h4>
                <p>{mission.description.substring(0, 100)}</p>
                <button type="button">Not a Member</button>
                <button type="button">Join Mission</button>
              </li>
            ))
        }
      </ul>

    </div>
  );
};

export default Mission;
