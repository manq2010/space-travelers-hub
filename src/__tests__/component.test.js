import { BrowserRouter, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Mission from '../features/Mission/Mission';
import AppRoutes from '../Routes';
import NotFound from '../features/NotFound/NotFound';
import MyProfilePage from '../pages/MyProfilePage';
import store from '../features/configureStore';
import Rocket from '../features/Rocket/rock';

// Routing test

describe('Navbar renders', () => {
  test('should redirect and update history', () => {
    const history = createMemoryHistory();
    render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <AppRoutes />
        </Router>
      </Provider>,
    );

    fireEvent.click(screen.getByTestId(/Missions/));
    expect(history.location.pathname).toEqual('/missions');
  });

  test('should redirect and update dom', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <AppRoutes />
        </BrowserRouter>
      </Provider>,
    );

    fireEvent.click(screen.getByTestId(/Missions/));
    expect(screen.getByText(/Description/)).toBeInTheDocument();
  });
});

describe('Not-found page is rendered', () => {
  test('Component is rendered', () => {
    render(
      <Provider store={store}>
        <NotFound />
      </Provider>,
    );
    const container = screen.getByText(/Oops Page/);
    expect(container).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <NotFound />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('HomePage component is rendererd', () => {
  test('Component is rendererd', () => {
    render(
      <Provider store={store}>
        <MyProfilePage />
      </Provider>,
    );
    const container = screen.getByText(/My Missions/);
    expect(container).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <MyProfilePage />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Mission component is rendererd', () => {
  test('Component is rendererd', () => {
    render(
      <Provider store={store}>
        <Mission />
      </Provider>,
    );
    const container = screen.getByText(/Description/);
    expect(container).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Mission />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Rockets component is rendererd', () => {
  test('Component is rendererd', () => {
    render(
      <Provider store={store}>
        <Rocket />
      </Provider>,
    );
    const container = screen.getByText(/Reserve Rocket/);
    expect(container).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Rocket />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});