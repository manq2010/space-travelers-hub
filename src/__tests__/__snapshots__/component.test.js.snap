// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage component is rendererd matches snapshot 1`] = `
<div>
  <div
    className="sc-dIfARi eOvQBd"
  >
    <div
      className="sc-hHTYSt iPKwjX"
    >
      <h2>
        My Missions
      </h2>
      <h3>
        no missions joined
      </h3>
      <div
        className="sc-dmctIk hkmTTW"
      />
    </div>
    <div
      className="sc-hHTYSt iPKwjX"
    >
      <h2>
        My Rockets
      </h2>
      <h3>
        no rockets reserved
      </h3>
      <div
        className="sc-dmctIk hkmTTW"
      />
    </div>
  </div>
</div>
`;

exports[`Mission component is rendererd matches snapshot 1`] = `
<div
  className="sc-bcXHqe bhFXJA"
>
  <ul
    className="sc-gswNZR hROBrh"
  >
    <li
      className="sc-dkrFOg eQyQvm"
    >
      <h4
        className="sc-hLBbgP sc-eDvSVe huPzVN ghLCpJ"
      >
        Mission
      </h4>
      <h4
        className="sc-hLBbgP sc-jSUZER huPzVN dMmJdR"
      >
        Description
      </h4>
      <h4
        className="sc-hLBbgP huPzVN"
      >
        Status
      </h4>
      <h4
        className="sc-hLBbgP huPzVN"
      />
    </li>
  </ul>
</div>
`;

exports[`Not-found page is rendered matches snapshot 1`] = `
<div
  className="sc-idXgbr cgkRCq"
>
  <h3>
    404!
  </h3>
  <p>
    Oops Page
     
    <em>
      Not
    </em>
     
    Found!
  </p>
</div>
`;
